internal/controlplane/handlers_repositories.go

```diff
@@ -401,60 +421,98 @@ func (s *Server) listRemoteRepositoriesForProvider(
 	return results, nil
 }
 
-// TODO: this probably can probably be used elsewhere
-// returns project ID and github client - this flow of code exists in multiple
-// places
-func (s *Server) getProviderAndClient(
+// covers the common logic for the two varieties of repo deletion
+func (s *Server) deleteRepository(
 	ctx context.Context,
-	projectID uuid.UUID,
-	request HasProtoContext,
-) (*db.Provider, v1.GitHub, error) {
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, request, projectID)
+	repoQueryMethod func() (db.Repository, error),
+) error {
+	projectID := getProjectID(ctx)
+
+	repo, err := repoQueryMethod()
+	if errors.Is(err, sql.ErrNoRows) {
+		return status.Errorf(codes.NotFound, "repository not found")
+	} else if err != nil {
+		return status.Errorf(codes.Internal, "unexpected error fetching repo: %v", err)
+	}
+
+	provider, err := s.findProviderByName(ctx, repo.Provider, projectID)
 	if err != nil {
-		return nil, nil, providerError(err)
+		return status.Errorf(codes.Internal, "cannot get provider: %v", err)
 	}
 
+	client, err := s.getClientForProvider(ctx, provider)
+	if err != nil {
+		return status.Errorf(codes.Internal, "cannot get client for provider: %v", err)
+	}
+
+	err = s.repos.DeleteRepository(ctx, client, &repo)
+	if err != nil {
+		return status.Errorf(codes.Internal, "unexpected error deleting repo: %v", err)
+	}
+	return nil
+}
+
+// inferProviderByOwner returns the provider to use for a given repo owner
+func (s *Server) inferProviderByOwner(ctx context.Context, owner string, projectID uuid.UUID, providerName string,
```

Comment:
@evankanderson This was something we had talked about. Do you have any ideas on how else we can do this?

Reply:
I think we want to do something like this:

```golang
opts, err := s.getProvidersByTrait(ctx, s.store, nil, projectID, db.ProviderTypeGithub)
if err != nil {
  return nil, err
}

slices.SortFunc(opts, func(a, b db.Provider) int {
  // Sort GitHub provider last
  if a.Class == db.ProviderClassGithub && b == db.ProviderClassGithubApp {
    return 1
  }
  if a.Class = db.ProviderClassGithubApp && b.Class == db.ProviderClassGitHub {
    return -1
  }
  return 0
})

for _, prov := range slices {
  if github.CanHandleOwner(ctx, prov, owner) {  // wrapper around checking if the creds can access the repo
    return prov, nil
  }
}

return nil, fmt.Errorf("No providers can handle repo owned by %s", owner)
```

And we could start with the name comparison for ProviderName you have here, but eventually switch to calling `ListHooks` or the like to make sure that we have admin permissions.